csharp.test/TestRowOrientedParquetFile.cs

```diff
@@ -456,6 +494,71 @@ private struct Row2
             public decimal D { get; set; }
         }
 
+        private sealed class Row3 : IEquatable<Row3>
+        {
+            public int A;
+            public VolumeInDollars B;
+            public bool Equals(Row3? other)
+            {
+                if (ReferenceEquals(null, other)) return false;
+                if (ReferenceEquals(this, other)) return true;
+                return A == other.A && B.Equals(other.B);
+            }
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private readonly struct VolumeInDollars : IEquatable<VolumeInDollars>
+        {
+            public VolumeInDollars(float value) { Value = value; }
+            public readonly float Value;
+            public bool Equals(VolumeInDollars other) => Value.Equals(other.Value);
+        }
+
+        /// <summary>
+        /// A read converter factory that supports our custom type.
+        /// </summary>
+        private sealed class ReadConverterFactory : LogicalReadConverterFactory
+        {
+            public override Delegate? GetDirectReader<TLogical, TPhysical>()
+            {
+                // Optional: the following is an optimisation and not stricly needed (but helps with speed).
+                // Since VolumeInDollars is bitwise identical to float, we can read the values in-place.
+                if (typeof(TLogical) == typeof(VolumeInDollars)) return LogicalRead.GetDirectReader<VolumeInDollars, float>();
+                return base.GetDirectReader<TLogical, TPhysical>();
+            }
+
+            public override Delegate GetConverter<TLogical, TPhysical>(ColumnDescriptor columnDescriptor, ColumnChunkMetaData columnChunkMetaData)
+            {
+                // VolumeInDollars is bitwise identical to float, so we can reuse the native converter.
+                if (typeof(TLogical) == typeof(VolumeInDollars)) return LogicalRead.GetNativeConverter<VolumeInDollars, float>();
+                return base.GetConverter<TLogical, TPhysical>(columnDescriptor, columnChunkMetaData);
+            }
+        }
+
+        /// <summary>
+        /// A logical type factory that supports our user custom type (for the write tests only). Rely on overrides (used by unit tests that can provide a columnLogicalTypeOverride).
+        /// </summary>
+        private sealed class WriteTypeFactory : LogicalTypeFactory
```

Comment:
Another issue is you use this type factory for both the read and write operations, but currently it only overrides `TryGetParquetTypes` which is used for writing. You'll need to implement `GetSystemTypes` for this to work for reading too.

Reply:
Oh, I didn't notice that, thank you!